agent/agentcontainers/api.go

```diff
@@ -1224,6 +1232,164 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
 	}, nil
 }
 
+func (api *API) devcontainerByIDLocked(devcontainerID string) (codersdk.WorkspaceAgentDevcontainer, error) {
+	for _, knownDC := range api.knownDevcontainers {
+		if knownDC.ID.String() == devcontainerID {
+			return knownDC, nil
+		}
+	}
+
+	return codersdk.WorkspaceAgentDevcontainer{}, httperror.NewResponseError(http.StatusNotFound, codersdk.Response{
+		Message: "Devcontainer not found.",
+		Detail:  fmt.Sprintf("Could not find devcontainer with ID: %q", devcontainerID),
+	})
+}
+
+func (api *API) handleDevcontainerDelete(w http.ResponseWriter, r *http.Request) {
+	var (
+		ctx            = r.Context()
+		devcontainerID = chi.URLParam(r, "devcontainer")
+	)
+
+	if devcontainerID == "" {
+		httpapi.Write(ctx, w, http.StatusBadRequest, codersdk.Response{
+			Message: "Missing devcontainer ID",
+			Detail:  "Devcontainer ID is required to delete a devcontainer.",
+		})
+		return
+	}
+
+	api.mu.Lock()
+
+	dc, err := api.devcontainerByIDLocked(devcontainerID)
+	if err != nil {
+		api.mu.Unlock()
+		httperror.WriteResponseError(ctx, w, err)
+		return
+	}
+
+	// NOTE(DanielleMaywood):
+	// We currently do not support canceling the startup of a dev container.
+	if dc.Status == codersdk.WorkspaceAgentDevcontainerStatusStarting {
```

Comment:
suggestion: Would it make sense to define a method `IsTerminal()` on the `WorkspaceAgentDevcontainerStatus` type that returns true for any *-ing state? My understanding is that we don't want to allow any modifications to *-ing devcontainers.